client/my-sites/plans-features-main/hooks/data-store/use-pricing-meta-for-grid-plans.ts

```diff
@@ -18,28 +22,88 @@ import type { IAppState } from 'calypso/state/types';
 interface Props {
 	planSlugs: PlanSlug[];
 	withoutProRatedCredits?: boolean;
+	storageAddOns?: ( AddOnMeta | null )[] | null;
+	currencyCode?: string | null;
+}
+
+/***
+ * The 2023 Pricing Grid handles prices differently than add-ons, so for the time being we need to
+ * do some conversion to calculate totals correctly.
+ *
+ * The 2023 pricing grid currently handles prices in some currencies in their smallest unit, like
+ * the Japanese Yen. For others, however, like USD, prices are not handled in their smallest unit,
+ * like the US dollar.
+ *
+ * For example, 2000 yen would be represented by the integer 2000 since a yen is the smallest unit.
+ * $20.00, however, would be represented as the integer 20, and not 2000. Since storage add-on
+ * prices are always handled in the smallest unit, we need to convert them, when appropriate, to the
+ * correct integer.
+ *
+ * We should expose a more friendly function from the format-currency package
+ */
+function convertPriceForSmallestUnit( price: number, currencyCode: string | null ) {
```

Comment:
I've tested the JPY case and it works. However, now that useAddOnDisplayCost() has been fixed by @sirbrillig in #81461, I'm not sure I understand why we'd still need to fix it this way. In JPY, I know https://wordpress.com/add-ons used to show the wrong price like you described in Automattic/martech#2151. However, now that it's showing the correct price, why we'd still need to convert it this way? 🤔

If we really have to, then I think we should check in with @Automattic/shilling to figure out a consistent price representation instead of hacking our way out for a more future proof solution :)

Reply:
Yes, I'm trying to grok the reason for this function and I'm not sure I do. The reason we are using an integer price in the smallest unit is that it prevents rounding errors when doing math (floating point math too inaccurate for prices). The only time we should need to convert it out of that format is when we display the price, which is what the `format-currency` package does.

Reply:
Hey y'all! Thanks for the feedback -- I'm happy this is generating good discussion. I'll try to clarify.

With the storage add-ons feature, there's a point where we need to calculate `add-on` prices + `plan` prices in a dynamic way. The arithmetic operation requires floats or integers, which, from what I understand, `format-currency` doesn't return. For example, $20 add-on + $20 plan = $40 total

Right now, `add-on` prices only return prices in the smallest unit ( Ex. USD $20 is represented in a value that corresponds to cents, so it's returned as the integer 2000 ).

`plan` prices, however, are _not_ handled in the smallest unit. ( Ex. USD $20 is represented as dollars, so it's handled as the integer 20 )

Fixing this representation mismatch before we handle the arithmetic of prices is what I was trying to address -- the closest exposed method I could find to help me with this was `getCurrencyObject`, which, internally, converts smallest unit integers.

I was planning to write a P2 this morning ( I couldn't yesterday, because I didn't see the bug until later at night 🙃 ), so I pushed something for discussion. With that being said, I see several ways to handle this. We could:

- expose a method from the @automattic/format-currency package to convert smallest unit integer prices
- return a price from the add-ons endpoint that _is not_ smallest unit
- longer term, we might look into consolidating representations of all prices in the 2023 plans grid into smallest units

Reply:
> plan prices, however, are not handled in the smallest unit. ( Ex. USD $20 is represented as dollars, so it's handled as the integer 20 )

Where do these plan prices come from? Presumably there's a wpcom endpoint somewhere that is returning them. If so, you can use the `Store_Price` class to transform the number easily into a smallest unit. 

I think that's the best solution if it's possible. We've done that many time for various endpoints. It carries the additional benefit that any math you do will be more accurate.

> return a price from the add-ons endpoint that is not smallest unit ( this seems like the best answer to me, but I'm not totally certain )

I would recommend not doing that because then other developers might use it for convenience, and we want to discourage using prices on the frontend that are not the smallest unit.

> expose a method from the @automattic/format-currency package to convert smallest unit integer prices

This would be ok, but I'd still like to avoid it if possible for the same reason as above: it would encourage other developers to convert integer prices to floats rather than the other way around, leading to more cases of incorrect rounding and more customer complaints.

Reply:
> expose a method from the @automattic/format-currency package to convert smallest unit integer prices

🤔 Actually I may have misunderstood this. Do you mean a method to convert a "common" unit to a smallest unit, or a smallest unit to a "common" unit? I think the first one would actually be ok, although still providing the smallest unit in the first place would be ideal.

Reply:
> Do you mean a method to convert a "common" unit to a smallest unit, or a smallest unit to a "common" unit? I think the first one would actually be ok, although still providing the smallest unit in the first place would be ideal.

I meant the second, and not the first, but your points about developer convenience doing more harm than good are compelling 👍